server/lib/config/nginx.conf

```diff
@@ -126,7 +126,17 @@ http {
             client_max_body_size     100G;
         }
 
-        location /dashboard {
+        location ~ ^/dashboard$ {
+            port_in_redirect off;
+
+            # If the URI omits the trailing slash, add it and reevaluate the
+            # location.  Note that this only happens if the URI does _not_
+            # begin with /dashboard/, as that will match the location below
+            # instead.
+            rewrite ^ /dashboard/ permanent;
+        }
+
+        location ^~ /dashboard/ {
```

Comment:
> So what exactly does the addition of the ^~ accomplish vs the old form?

Per the [docs](http://nginx.org/en/docs/http/ngx_http_core_module.html#location),
> If the longest matching prefix location has the `^~` modifier then regular expressions are not checked.

So, this prevents recursive rewrites:  after the rewrite, the `location` at line 139 matches and disables further match-searching, so the regex match is not reevaluated.  (This was critical in the previous iteration of this code; however, now that I've got the no-trailing-slash location anchored at both ends, we _could_ get by without this, but having it makes the search more efficient.)

> when we just load `/dashboard/`, are we loading `$uri/index.html` (line 143's `try_files`), which is `/dashboard//index.html`, or are we loading the `/dashboard/index.html` literal fallback?

I expect that we try to load `/dashboard/` which fails because it's a directory; then we probably try to load `/dashboard//index.html` which probably succeeds (Nginx may squash out the extra slash, and, if not, then `open()` call probably ignores it); and, if that fails, we load the fallback, `/dashboard/index.html`.